trace/trace_test.go

```diff
@@ -437,3 +446,303 @@ func TestSpanContextFromContext(t *testing.T) {
+				kvs: []label.KeyValue{
+					label.String("key1", "val1"),
+					label.String("key2", "val2"),
+					label.String("key3", "val3"),
+				},
+			},
+			key: "keyx",
+			expectedTraceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key1", "val1"),
+					label.String("key2", "val2"),
+					label.String("key3", "val3"),
+				},
+			},
+		},
+		{
+			name: "Invalid key",
+			traceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key1", "val1"),
+					label.String("key2", "val2"),
+					label.String("key3", "val3"),
+				},
+			},
+			key: "in va lid",
+			expectedTraceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key1", "val1"),
+					label.String("key2", "val2"),
+					label.String("key3", "val3"),
+				},
+			},
+			expectedErr: errInvalidTraceStateKeyValue,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			result, err := tc.traceState.Delete(tc.key)
+			if tc.expectedErr != nil {
+				require.Error(t, err)
+				assert.Equal(t, tc.expectedErr, err)
+				assert.Equal(t, tc.traceState, result)
+			} else {
+				require.NoError(t, err)
+				assert.Equal(t, tc.expectedTraceState, result)
+			}
+		})
+	}
+}
+
+func TestTraceStateInsert(t *testing.T) {
+	testCases := []struct {
+		name               string
+		traceState         TraceState
+		keyValue           label.KeyValue
+		expectedTraceState TraceState
+		expectedErr        error
+	}{
+		{
+			name: "OK case - add new",
+			traceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key1", "val1"),
+					label.String("key2", "val2"),
+					label.String("key3", "val3"),
+				},
+			},
+			keyValue: label.String("key4@vendor", "val4"),
+			expectedTraceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key4@vendor", "val4"),
+					label.String("key1", "val1"),
+					label.String("key2", "val2"),
+					label.String("key3", "val3"),
+				},
+			},
+		},
+		{
+			name: "OK case - replace",
+			traceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key1", "val1"),
+					label.String("key2", "val2"),
+					label.String("key3", "val3"),
+				},
+			},
+			keyValue: label.String("key2", "valX"),
+			expectedTraceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key2", "valX"),
+					label.String("key1", "val1"),
+					label.String("key3", "val3"),
+				},
+			},
+		},
+		{
+			name: "Invalid key/value",
+			traceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key1", "val1"),
+				},
+			},
+			keyValue: label.String("key!", "val!"),
+			expectedTraceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key1", "val1"),
+				},
+			},
+			expectedErr: errInvalidTraceStateKeyValue,
+		},
+		{
+			name:               "Too many entries",
+			traceState:         TraceState{generateKVsWithMaxMembers()},
+			keyValue:           label.String("keyx", "valx"),
+			expectedTraceState: TraceState{generateKVsWithMaxMembers()},
```

Comment:
I feel like this should take into consideration the [truncation scheme of the tracestate](https://www.w3.org/TR/trace-context-1/#tracestate-limits):

> In a situation where tracestate needs to be truncated due to size limitations, the vendor MUST truncate whole entries. Entries larger than 128 characters long SHOULD be removed first. Then entries SHOULD be removed starting from the end of tracestate.

This would mean that `keyx=valux` would be added to the beginning and the last entry would be dropped based on my understanding. I think returning an error as well as implementing this logic seems appropriate.

Reply:
My understanding of the specification was this applies only to 'size limitations', considering the context of the section (https://www.w3.org/TR/trace-context-1/#tracestate-limits). i.e. situations where it is necessary to truncate the header due to length and not due to number of members in the list. However, now I'm unsure what would be the expected behavior with regards to exceeding max number of members.

Reply:
That's a good point. I'm not entirely sure what behavior is expected either now that you point this out. I'm going to ask around with people who developed that specification to see if I can get clarity.

Reply:
Opened https://github.com/w3c/trace-context/issues/440

Reply:
Same here regarding resolving this and I'll open an follow on issue/PR when the w3c issue is resolved.